In [1]:
import numpy as np
import pandas as pd
df = pd.read_csv("RewardsData.csv")


In [2]:
df.isnull()

,User ID,Birthdate,City,State,Zip,Available Points,Total Points Earned,Points Spent,Joined On,Last Seen,Tags
0,False,False,False,False,False,False,False,False,False,False,False
1,False,True,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...
7247,False,True,True,True,True,False,False,False,False,False,False
7248,False,True,True,True,True,False,False,False,False,False,False
7249,False,False,True,True,True,False,False,False,False,False,False
7250,False,True,True,True,True,False,False,False,False,False,False


In [4]:
df.isnull().sum()

User ID                   0
Birthdate              6080
City                   6922
State                  6922
Zip                    6922
Available Points          0
Total Points Earned       0
Points Spent              0
Joined On                 0
Last Seen               221
Tags                      0
dtype: int64

In [6]:
df.isnull().mean()*100

User ID                 0.000000
Birthdate              83.838941
City                   95.449531
State                  95.449531
Zip                    95.449531
Available Points        0.000000
Total Points Earned     0.000000
Points Spent            0.000000
Joined On               0.000000
Last Seen               3.047435
Tags                    0.000000
dtype: float64

In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7252 entries, 0 to 7251
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   User ID              7252 non-null   int64
 1   Birthdate            1172 non-null   str  
 2   City                 330 non-null    str  
 3   State                330 non-null    str  
 4   Zip                  330 non-null    str  
 5   Available Points     7252 non-null   int64
 6   Total Points Earned  7252 non-null   int64
 7   Points Spent         7252 non-null   int64
 8   Joined On            7252 non-null   str  
 9   Last Seen            7031 non-null   str  
 10  Tags                 7252 non-null   str  
dtypes: int64(4), str(7)
memory usage: 623.3 KB


In [10]:
df["Last Seen"]=df["Last Seen"].fillna(df["Joined On"])

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7252 entries, 0 to 7251
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   User ID              7252 non-null   int64
 1   Birthdate            1172 non-null   str  
 2   City                 330 non-null    str  
 3   State                330 non-null    str  
 4   Zip                  330 non-null    str  
 5   Available Points     7252 non-null   int64
 6   Total Points Earned  7252 non-null   int64
 7   Points Spent         7252 non-null   int64
 8   Joined On            7252 non-null   str  
 9   Last Seen            7252 non-null   str  
 10  Tags                 7252 non-null   str  
dtypes: int64(4), str(7)
memory usage: 623.3 KB


In [12]:
df = df.dropna(subset=["Last Seen"])

In [14]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7252 entries, 0 to 7251
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   User ID              7252 non-null   int64
 1   Birthdate            1172 non-null   str  
 2   City                 330 non-null    str  
 3   State                330 non-null    str  
 4   Zip                  330 non-null    str  
 5   Available Points     7252 non-null   int64
 6   Total Points Earned  7252 non-null   int64
 7   Points Spent         7252 non-null   int64
 8   Joined On            7252 non-null   str  
 9   Last Seen            7252 non-null   str  
 10  Tags                 7252 non-null   str  
dtypes: int64(4), str(7)
memory usage: 623.3 KB


In [15]:
df.isnull().mean()*100

User ID                 0.000000
Birthdate              83.838941
City                   95.449531
State                  95.449531
Zip                    95.449531
Available Points        0.000000
Total Points Earned     0.000000
Points Spent            0.000000
Joined On               0.000000
Last Seen               0.000000
Tags                    0.000000
dtype: float64

Birthdate 6080 missing, dropping I'll lose thousands of rows.
better Keep it and fill missing values.

In [16]:
df["Birthdate"] = df["Birthdate"].fillna("Unknown")

In [17]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7252 entries, 0 to 7251
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   User ID              7252 non-null   int64
 1   Birthdate            7252 non-null   str  
 2   City                 330 non-null    str  
 3   State                330 non-null    str  
 4   Zip                  330 non-null    str  
 5   Available Points     7252 non-null   int64
 6   Total Points Earned  7252 non-null   int64
 7   Points Spent         7252 non-null   int64
 8   Joined On            7252 non-null   str  
 9   Last Seen            7252 non-null   str  
 10  Tags                 7252 non-null   str  
dtypes: int64(4), str(7)
memory usage: 623.3 KB


In [18]:
df["City"] = df["City"].fillna("Unknown")
df["State"] = df["State"].fillna("Unknown")
df["Zip"] = df["Zip"].fillna("Unknown")

In [20]:
df.shape

(7252, 11)

In [21]:
df.isnull()

,User ID,Birthdate,City,State,Zip,Available Points,Total Points Earned,Points Spent,Joined On,Last Seen,Tags
0,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...
7247,False,False,False,False,False,False,False,False,False,False,False
7248,False,False,False,False,False,False,False,False,False,False,False
7249,False,False,False,False,False,False,False,False,False,False,False
7250,False,False,False,False,False,False,False,False,False,False,False


In [22]:
df.isnull().mean()*100


User ID                0.0
Birthdate              0.0
City                   0.0
State                  0.0
Zip                    0.0
Available Points       0.0
Total Points Earned    0.0
Points Spent           0.0
Joined On              0.0
Last Seen              0.0
Tags                   0.0
dtype: float64